### PR TITLE
Feature/session api

### PR DESCRIPTION
**Pull request by truemped**
Again, sorry for all this spam!

So the final note: I am unsure about the session callback and if it is ok, the way it is? Any thoughts?

Best
Daniel

### CODE REVIEW

File: trombi/client.py [hunk 1]
```diff
@@ -257,6 +257,54 @@ class Server(TrombiObject):
...
+            if response.code in [200, 302] and store_cookie:
+                # store the cookie for subsequent requests with trombi
+                if cookie:
+                    if 'headers' not in self._fetch_args:
+                        self._fetch_args['headers'] = HTTPHeaders()
+                    h = self._fetch_args['headers']
+                    h['Cookie'] = cookie
+                    h['X-Couchdb-Www-Authenticate'] = "Cookie"
+
+            if response.request.method == "DELETE":
```
Comment (akheron) [inline trombi/client.py]: `if logout:` would be cleaner here

File: trombi/client.py [hunk 2]
```diff
@@ -876,4 +924,4 @@ class Paginator(TrombiObject):
         self._db.view(design_doc, viewname, _really_callback, **kwargs)
 
 
-VALID_DB_NAME = re.compile(r'^[a-z][a-z0-9_$()+-/]*$')
+VALID_DB_NAME = re.compile(r'^_?[a-z][a-z0-9_$()+-/]*$')
```
Comment (akheron) [inline trombi/client.py:927]: Why did you change the regexp? Names starting with an underscore are not valid for "normal" databases.

Comment (truemped) [inline trombi/client.py:927]: At least in my test case I need to sign a user up. And then I need access to the '_users' database... Should I also create a method for signup?

File: trombi/client.py (hunk 1 above)
Comment (truemped) [inline trombi/client.py]: Indeed! Will change that!

File: trombi/client.py (hunk 2 above)
Comment (nailor) [inline trombi/client.py:927]: I second @akheron here. If you need the access in your tests, you can just use direct HTTP calls. Signup method is ok to me, if that feels more convenient to you (and as long it is tested).

**Comment by nailor** [comment]
Thanks for the pull request and sorry that it took a while from me to get into this.

I commented on the VALID_DB_NAME inline, that's a change that I would not like to take in to the trombi. Otherwise the pull request is good in overall, but it needs a few changes before I'll apply it:
- As stated, the VALID_DB_NAME should be left as it was
- The license in test_session seems to be Apache license, but the project is licensed with MIT. I won't accept different licenses in the source tree, but feel free to change it to a MIT license under you name.

Overall a good quality pull request and the session support is a nice addition to trombi! I'm looking forward merging this with proposed changes so that I could release 1.0 version of trombi.

I'll close this pull request, but you can either do a new one or do additions in this.

File: trombi/client.py (hunk 2 above)
Comment (truemped) [inline trombi/client.py:927]: Well, I not only need it in my tests. Basically I want to be able to add users in CouchDB and use it for authentication of users on my site.

Maybe I should add a separate API for adding/removing/updating Users and use direct HTTP calls?

**Comment by truemped** [comment]
I will start to work on that again but first need a feedback if it would be nice to add the add/remove/update user stuff into the `trombi.Server` itself!

License header of the test_session file is wrong, sorry about that... My default vim template :) Will be changed!

File: trombi/client.py (hunk 2 above)
Comment (nailor) [inline trombi/client.py:927]: I'd go for a separate API for users, maybe some sort of a subclass of `trombi.Database`. I'm not comfortable of mixing editing users and "normal" databases, even though under the hood they are the same thing.

**Comment by nailor** [comment]
User stuff in `trombi.Server` is ok to me, but I'd leave it outside this pull request, so that we get the preliminary session support now. I'm more than willing to wait for a user API before pushing out 1.0!
